pkg/operator/controllers/previewfeature/nsgflowlogs/nsgflowlogs_test.go

```diff
@@ -0,0 +1,228 @@
+				// enable only once
+				client.EXPECT().CreateOrUpdateAndWait(gomock.Any(), networkWatcherResourceGroupName, networkWatcherName, subnetNameMasterNSGName, *flowLog)
+			},
+			instance: func(feature *aropreviewv1alpha1.PreviewFeature) {
+				feature.Spec.NSGFlowLogs.Enabled = true
+				feature.Spec.NSGFlowLogs.NetworkWatcherID = networkWatcherResourceId
+			},
+			wantErr: nil,
+		},
+		{
+			name: "disable flow log",
+			subnetMock: func(mock *mock_subnet.MockManager, kmock *mock_subnet.MockKubeManager) {
+				kmock.EXPECT().List(gomock.Any()).Return([]subnet.Subnet{
+					{
+						ResourceID: resourceIdMaster,
+					},
+					{
+						ResourceID: resourceIdWorker,
+					},
+				}, nil)
+				mock.EXPECT().Get(gomock.Any(), gomock.Any()).Return(&mgmtnetwork.Subnet{
+					SubnetPropertiesFormat: &mgmtnetwork.SubnetPropertiesFormat{
+						NetworkSecurityGroup: &mgmtnetwork.SecurityGroup{
+							ID: &subnetNameMasterNSGID,
+						},
+					},
+				}, nil)
+				mock.EXPECT().Get(gomock.Any(), gomock.Any()).Return(&mgmtnetwork.Subnet{
+					SubnetPropertiesFormat: &mgmtnetwork.SubnetPropertiesFormat{
+						NetworkSecurityGroup: &mgmtnetwork.SecurityGroup{
+							ID: &subnetNameMasterNSGID, // this intentionally has the same NSG ID as the master subnet. master and worker subnet share the same NSG as they do in a default ARO installation
+						},
+					},
+				}, nil)
+			},
+			flowLogClientMock: func(client *mock_network.MockFlowLogsClient) {
+				client.EXPECT().DeleteAndWait(gomock.Any(), networkWatcherResourceGroupName, networkWatcherName, subnetNameMasterNSGName)
+			},
+			instance: func(feature *aropreviewv1alpha1.PreviewFeature) {
+				feature.Spec.NSGFlowLogs.Enabled = false
+				feature.Spec.NSGFlowLogs.NetworkWatcherID = networkWatcherResourceId
+			},
+			wantErr: nil,
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			controller := gomock.NewController(t)
+			defer controller.Finish()
+
+			subnets := mock_subnet.NewMockManager(controller)
+			kubeSubnets := mock_subnet.NewMockKubeManager(controller)
+			if tt.subnetMock != nil {
+				tt.subnetMock(subnets, kubeSubnets)
+			}
+
+			instance := getValidPreviewFeatureInstance()
+			if tt.instance != nil {
+				tt.instance(instance)
+			}
+
+			flowLogsClient := mock_network.NewMockFlowLogsClient(controller)
+			if tt.flowLogClientMock != nil {
+				tt.flowLogClientMock(flowLogsClient)
+			}
+
+			r := NewFeature(flowLogsClient, kubeSubnets, subnets, location)
+
+			err := r.Reconcile(context.Background(), instance)
+			if err != nil {
+				if tt.wantErr == nil {
+					t.Fatal(err)
+				}
+				if err != nil && err.Error() != tt.wantErr.Error() || err == nil && tt.wantErr != nil {
+					t.Errorf("Expected Error %s, got %s when processing %s testcase", tt.wantErr.Error(), err.Error(), tt.name)
+				}
+			}
```

Comment:
I think with this we can miss cases when we want error, but `r.Reconcile(...)` wrongfully returns `nil`.

We can fix it by making `wantErr` to be string and doing this check instead:

```suggestion
			if err != nil && err.Error() != tt.wantErr ||
				err == nil && tt.wantErr != "" {
				t.Error(err)
			}
```